source/async_stream.c

```diff
@@ -0,0 +1,209 @@
+    job->my_future = aws_future_bool_acquire(future);
+
+    /* Kick off work  */
+    s_async_stream_fill_job_loop(job);
+
+    return future;
+}
+
+/*******************************************************************************
+ * aws_async_stream_wrapping_synchronous
+ ******************************************************************************/
+
+struct aws_async_stream_wrapping_synchronous {
+    struct aws_async_input_stream base;
+    struct aws_input_stream *source;
+};
+
+static void s_async_wrapping_synchronous_stream_destroy(struct aws_async_input_stream *async_stream) {
+    struct aws_async_stream_wrapping_synchronous *async_impl = async_stream->impl;
+    aws_input_stream_release(async_impl->source);
+    aws_mem_release(async_stream->alloc, async_impl);
+}
+
+static struct aws_future_bool *s_async_wrapping_synchronous_stream_read(
+    struct aws_async_input_stream *async_stream,
+    struct aws_byte_buf *dest) {
+
+    struct aws_async_stream_wrapping_synchronous *async_impl = async_stream->impl;
+
+    struct aws_future_bool *future = aws_future_bool_new(async_stream->alloc);
+
+    /* Keep calling read() until we get some data, or hit EOF.
+     * We do this because the synchronous aws_input_stream API allows
+     * 0 byte reads, but the aws_async_input_stream API does not.
+     *
+     * The synchronous aws_input_stream API allows 0 bytes reads because we
+     * didn't used to have an async API, and 0 byte reads were the way to report
+     * "data not available yet".
+     *
+     * Continually calling read() is a waste of CPU, but I'm not sure tricks to
+     * work around this (sleeping between retries, retrying from a thread, etc)
+     * are worth the complexity. The real solution to an aws_input_stream
+     * doing 0 byte reads is to replace it with an actual async stream */
+    size_t prev_len = dest->len;
+    struct aws_stream_status status = {.is_end_of_stream = false, .is_valid = true};
+    while (!status.is_end_of_stream && (dest->len == prev_len)) {
+        /* read from stream */
+        if (aws_input_stream_read(async_impl->source, dest) != AWS_OP_SUCCESS) {
+            aws_future_bool_set_error(future, aws_last_error());
+            goto done;
+        }
+
+        /* check if stream is done */
+        if (aws_input_stream_get_status(async_impl->source, &status) != AWS_OP_SUCCESS) {
+            aws_future_bool_set_error(future, aws_last_error());
+            goto done;
+        }
+    }
+
+    aws_future_bool_set_result(future, status.is_end_of_stream);
+done:
+    return future;
+}
+
+static const struct aws_async_input_stream_vtable s_async_stream_wrapping_input_stream_vtable = {
+    .destroy = s_async_wrapping_synchronous_stream_destroy,
+    .read = s_async_wrapping_synchronous_stream_read,
+};
+
+struct aws_async_input_stream *aws_async_input_stream_new_from_synchronous(
+    struct aws_allocator *alloc,
+    struct aws_input_stream *source) {
+
+    AWS_PRECONDITION(source);
```

Comment:
nit: precondition usage seems to be very sporadic. if we are doing it then it should be consistent for all nullable inputs

Reply:
consistently using it on all args to public functions now
not bothering to check in internal functions